src/child.rs

```diff
@@ -0,0 +1,176 @@
+/// to our stdout and stderr so that the user has an idea of what is going on
+/// behind the scenes.
+pub fn run(
+    logger: &Logger,
+    mut command: process::Command,
+    command_name: &str,
+) -> Result<String, failure::Error> {
+    info!(logger, "Running {:?}", command);
+
+    let mut child = command
+        .stdout(process::Stdio::piped())
+        .stderr(process::Stdio::piped())
+        .spawn()?;
+
+    let since = time::Instant::now();
+
+    let stdout = io::BufReader::new(child.stdout.take().unwrap());
+    let stderr = io::BufReader::new(child.stderr.take().unwrap());
+
+    let (send, recv) = mpsc::channel();
+    let stdout_send = send.clone();
+    let stderr_send = send;
+
+    // Because pipes have a fixed-size buffer, we need to keep reading stdout
+    // and stderr on a separate thread to avoid potential dead locks with
+    // waiting on the child process.
+
+    let stdout_handle = thread::spawn(move || {
+        for line in stdout.lines() {
+            stdout_send.send(line.map(Output::Stdout)).unwrap();
+        }
+    });
+
+    let stderr_handle = thread::spawn(move || {
+        for line in stderr.lines() {
+            stderr_send.send(line.map(Output::Stderr)).unwrap();
+        }
+    });
+
+    let mut stdout = String::new();
+    let mut stderr = String::new();
+    let mut is_long_running = false;
+
+    loop {
+        if !is_long_running && taking_too_long(since) {
+            // The command has now been taking too long. Print the buffered stdout and
+            // stderr, and then continue waiting on the child.
+            stdout
+                .lines()
+                .map(|l| Ok(Output::Stdout(l)))
+                .chain(stderr.lines().map(|l| Ok(Output::Stderr(l))))
+                .for_each(|l| print_child_output(l, command_name));
+
+            is_long_running = true;
+        }
+
+        // Get any output that's been sent on the channel without blocking.
+        handle_output(
+            recv.try_iter(),
+            logger,
+            is_long_running,
+            &mut stdout,
+            &mut stderr,
+            command_name,
+        );
+
+        if let Some(exit) = child.try_wait()? {
+            // Block on collecting the rest of the child's output.
+            handle_output(
+                recv,
+                logger,
+                is_long_running,
+                &mut stdout,
+                &mut stderr,
+                command_name,
+            );
+
+            // Join the threads reading the child's output to make sure the
+            // finish OK.
+            stdout_handle.join().unwrap();
+            stderr_handle.join().unwrap();
+
+            if exit.success() {
+                return Ok(stdout);
+            } else {
+                let msg = format!("`{}` did not exit successfully", command_name);
+                return Err(Error::cli(&msg, stderr.into(), exit).into());
```

Comment:
This'll want to ferry along stdout as well, want to be sure to not lose anything whenever we print failure messages!

I think it'd also be good to carry along the full command line here as opposed to just the name of the command to ensure that all the arguments are in the error message too